packages/cli/src/lib/packager/packageManager.ts

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright 2023 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { join as joinPath } from 'path';
+import { existsSync } from 'fs';
+import { findRootSync } from '@manypkg/find-root';
+
+interface PackageManager {
+  command: string;
+  lockfile: string;
+}
+
+function detect(): PackageManager {
+  const rootDir = findRootSync(process.cwd());
+  const lockfiles = {
```

Comment:
Would appreciate an entry for `pnpm` here to close issue #20658 with the code:
`'pnpm-lock.yaml': 'pnpm',`

Are the lockfiles actually required elsewhere in the code? Or, are they simply used to detect the package manager? If the latter, wouldn't be easier to standardize on the "packageManager" entry in the "package.json" file and require it? The lockfile would not be necessary unless it is used elsewhere in the code. If it is necessary, then a code comment above this variable should indicate that all package managers and their respective lockfiles must be added to this array. What happens if another package manager uses the same lockfile format? For example, the `cnpm` mirror for China uses "package-lock.json" and would conflict with standard `npm` (the first entry in the array would always win).

Reply:
Actually @ourchitectureio could you test with pnpm?

How:
- Check out my PR branch / fork
- Add the pnpm entry
- cd packages/cli ; yarn build
- Then in _your_ backstage backend: use the local `@backstage/cli` . In my case i do this with `npm i -D ~/backstage-fork/packages/cli`. Then run `build:backend`

Reply:
I'll take a look at this starting tomorrow (Wednesday, ET)

Reply:
This got really deep, really quick. The CLI makes use of `yarn` commands in multiple places and parses the `yarn.lock` file in very specific ways. In order to do this properly, the "Lockfile" implementation has to be completely refactored to support a strategy pattern for different package managers, like `npm`, `yarn`, `pnpm`, etc.

This includes a strategy for parsing different lock files:

- `yarn` and the existing ["yarn.lock"](https://www.npmjs.com/package/@yarnpkg/parsers)
- `npm` and ["package-lock.json"](https://github.com/snyk/nodejs-lockfile-parser)
- `pnpm` and ["pnpm-lock.yaml"](https://www.npmjs.com/package/@pnpm/lockfile-file)

Additionally, there are assumptions about `yarn` workspaces, like "paths" `getTargetRoot` function which checks for the "package.json" contents to have a `.workspaces.packages` entry (`yarn`-specific). Something like `pnpm` would either need a check for a "pnpm-workspace.yaml" file, but... this code isn't looking for a potential mono-repo target root, but rather a Backstage project root. In a monorepo, the workspaces could be defined at an even higher level. So, a robust solution would likely require a new backstage-specific config entry like `.backstage.isRoot: true` instead of relying on the `yarn` workspace configuration.

Reply:
Thanks so much for trying it with pnpm @ourchitectureio . For your use case, do you need anything else than build? E.g. i only need build:backend which runs `backstage-cli package build` inside packages/backend.

The workspace aspect is not strictly required from what i can see for that use case. You could try again with the first commit of this PR (`480585a3ad30333be1e91923b2c8dab214610c63`) that does not rely on "paths" but https://www.npmjs.com/package/@manypkg/find-root which says it supports pnpm too.

Reply:
Yes. We would want to see end-to-end support for alternative package managers across all commands.